Dotnet mattkeenum

MattKeenum opened the pull request:
I've made several modifications to the dotnet bindings. 

Notable changes:
- I've added a Prerun class to help constrain how preruns should be initialized 
- I added a test to ensure the new Prerun is working as expected
- Updated the endpoint for the west data center
- Updated EdgeVersions to match what Sauce Labs is currently offering
- Updated Safari Versions/Platforms to match what Sauce Labs is currently offering
- Added several overloaded constructors for the SauceSession class
- Simplified the SauceSession class to be more DRY
- Finished the Sauce Driver class
- Several refactorings across other classes

MattKeenum committed:
Major modifications

I made changes to most files, providing full functionality.

MattKeenum committed:
Added test for PreRun class

Removed need for Prerun to be serialized as a dictionary and removed the need for timespan.

Corrected the Baseline tests to have public driver and Test Context

nadvolod commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- dotnet/Simple.Sauce/IncorrectSafariVersionException.cs  1
- dotnet/Simple.Sauce/JsonUtils.cs  1
- dotnet/Simple.Sauce/SauceDriver.cs  1
- dotnet/Simple.Sauce/SauceOptions.cs  2
- dotnet/SimpleSauce.Test/PreRunTests.cs  1
- dotnet/Simple.Sauce/PreRun.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844&bid=20543758)

nadvolod commented on dotnet/SimpleSauce.Test/BaselineSeleniumAcceptanceTests.cs:
```diff
@@ -13,33 +14,47 @@ namespace SauceBindings.Test
-            driver.Navigate().GoToUrl("https://www.google.com");
-            driver.SessionId.Should().NotBeNull();
+            _driver = new RemoteWebDriver(new Uri("https://ondemand.us-west-1.saucelabs.com/wd/hub"),
+                chromeOptions.ToCapabilities(), TimeSpan.FromSeconds(30)); ;
```
Codacy found an issue: [Remove this empty statement.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844)

nadvolod commented on dotnet/SimpleSauce.Test/SauceOptionsTests.cs:
```diff
@@ -62,19 +63,20 @@ public void AcceptsAllW3CValues()
             var args = new List<string> { "--silent", "-a", "-q" };
 
-            var prerun = new Dictionary<string, object>
+            PreRun prerun = new PreRun()
```
Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844)

nadvolod commented on dotnet/Simple.Sauce/SauceOptions.cs:
```diff
@@ -55,83 +59,114 @@ public SauceOptions()
         }
 
-        public void WithSafari()
+        public SauceOptions WithFirefox(string version = DEFAULT_BROWSER_VERSION, string platform = DEFAULT_PLATFORM)
```
Codacy found an issue: [Use the overloading mechanism instead of the optional parameters.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844)

nadvolod commented on dotnet/SimpleSauce.Test/PreRunTests.cs:
```diff
@@ -0,0 +1,41 @@
+        [TestMethod]
+        public void PreRunTest()
+        {
+            SauceOptions.Prerun = new PreRun()
```
Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844)

nadvolod commented on dotnet/Simple.Sauce/IncorrectSafariVersionException.cs:
```diff
@@ -4,5 +4,7 @@ namespace Sauce.Bindings
 {
     public class IncorrectSafariVersionException : Exception
     {
+        public IncorrectSafariVersionException() : base() { }
```
Codacy found an issue: [Remove this redundant 'base()' call.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844)

nadvolod commented on dotnet/Simple.Sauce/SauceOptions.cs:
```diff
@@ -55,83 +59,114 @@ public SauceOptions()
         }
 
-        public void WithSafari(string safariVersion)
+        public SauceOptions WithSafari(string version = DEFAULT_BROWSER_VERSION)
```
Codacy found an issue: [Use the overloading mechanism instead of the optional parameters.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844)

nadvolod commented on dotnet/Simple.Sauce/SauceOptions.cs:
```diff
@@ -55,83 +59,114 @@ public SauceOptions()
         }
 
-        public void WithChrome(string chromeVersion)
+        public SauceOptions WithChrome(string version = DEFAULT_BROWSER_VERSION, string platform = DEFAULT_PLATFORM)
```
Codacy found an issue: [Use the overloading mechanism instead of the optional parameters.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844)

MattKeenum committed:
Updated per Codacy

IncorrectSafariVersionException: removed redundant base call
SauceOptions: Removed optional parameters and created overloaded methods
BaselineSeleniumAcceptanceTests: Removed unnesseccary semicolon
PreRunTests: removed redundant paranthesis
SauceOptionsTests: removed redundant paranthesis

nadvolod commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dotnet/SimpleSauce.Test/PreRunTests.cs  1
- dotnet/Simple.Sauce/JsonUtils.cs  1
- dotnet/Simple.Sauce/PreRun.cs  1
- dotnet/Simple.Sauce/IncorrectSafariVersionException.cs  1
- dotnet/Simple.Sauce/SauceOptions.cs  2
- dotnet/Simple.Sauce/SauceDriver.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6319844&bid=20543758)

nadvolod commented:
Excellent work here @MattKeenum 👏 This is a huge PR. Can you please break it up into a few smaller PRs that we can address one at a time. Otherwise, we might be dealing with this code review for a while 😨

MattKeenum commented:
> Excellent work here @MattKeenum 👏 This is a huge PR. Can you please break it up into a few smaller PRs that we can address one at a time. Otherwise, we might be dealing with this code review for a while 😨

No problem. Is there a specific way you want me to break it up? Is per file the best? Or more per module?

nadvolod commented:
@MattKeenum smaller is better 👍  here are a few super easy ones:
- dotnet/Simple.Sauce/Browser.cs
- dotnet/Simple.Sauce/DataCenter.cs
- dotnet/Simple.Sauce/EdgeVersion.cs
- dotnet/Simple.Sauce/Platforms.cs
See if you can find a few more simple ones like that. Afterwards, it gets more hairy as we're working on a large update to the API. So I'll need to work and coordinate with you to figure that out

MattKeenum commented:
> @MattKeenum smaller is better 👍 here are a few super easy ones:
> 
> * dotnet/Simple.Sauce/Browser.cs
> * dotnet/Simple.Sauce/DataCenter.cs
> * dotnet/Simple.Sauce/EdgeVersion.cs
> * dotnet/Simple.Sauce/Platforms.cs
>   See if you can find a few more simple ones like that. Afterwards, it gets more hairy as we're working on a large update to the API. So I'll need to work and coordinate with you to figure that out

So do you want me to break these down as individual pull requests?
